File: src/Layer.cpp
#include "../include/Layer.h"
#include <algorithm>
#include <stdexcept>
void Layer::loadFrames(pugi::xml_node& layerNode) noexcept {
	auto frames = layerNode.child("frames").children("DOMFrame");
	for (auto iter = frames.begin(); iter != frames.end(); ++iter) {
		this->frames.push_back(std::make_unique<Frame>(*iter));
	}
}
Layer::Layer(pugi::xml_node& layerNode) noexcept {
	this->root = layerNode;
	this->setColor(layerNode.attribute("color").empty() ? "#000000" : layerNode.attribute("color").value());
	this->setLayerType(layerNode.attribute("layerType").empty() ? "normal" : layerNode.attribute("layerType").value());
	this->setLocked(layerNode.attribute("locked").as_bool());
	this->setName(layerNode.attribute("name").empty() ? "layer" : layerNode.attribute("name").value());
	this->setParentLayerIndex((layerNode.attribute("parentLayerIndex").empty()) ? std::nullopt : std::make_optional<unsigned int>(layerNode.attribute("parentLayerIndex").as_uint()));
	loadFrames(this->root);
}
Layer::~Layer() noexcept {

}

void Layer::removeKeyframe(unsigned int keyFrameIndex) noexcept {
	Frame* frame = this->frames[keyFrameIndex].get();
	this->getRoot().child("frames").remove_child(frame->getRoot());
	this->frames.erase(this->frames.begin() + keyFrameIndex);
}

bool Layer::clearKeyFrame(unsigned int frameIndex) noexcept {
	unsigned int index = this->getKeyframeIndex(frameIndex);
	Frame* frame = this->frames[index].get();
	if (frameIndex != frame->getStartFrame()) return false;
	// Special case for if there's one keyframe: replace it with a blank one (remove all elements)
	if (this->frames.size() == 1) {
		if(frame->isEmpty()) return false;
		frame->clearElements();
		return true;
	}
	// Special case for the first keyframe: replace it with the next one
	if(index == 0) {
		Frame* nextFrame = this->frames[index + 1].get();
		nextFrame->setDuration(nextFrame->getDuration() + frame->getDuration());
		nextFrame->setStartFrame(index);
		this->removeKeyframe(index);
		return true;
	}
	Frame* prevFrame = this->frames[index - 1].get();
	prevFrame->setDuration(prevFrame->getDuration() + frame->getDuration());
	this->removeKeyframe(index);
	return true;
}

bool Layer::insertKeyframe(unsigned int frameIndex, bool isBlank) noexcept {
	unsigned int index = this->getKeyframeIndex(frameIndex);
	Frame* frame = this->frames[index].get();
	if (frameIndex == frame->getStartFrame()) {
		frameIndex++;
		if (frameIndex >= this->getFrameCount()) return false;
		unsigned int newIndex = this->getKeyframeIndex(frameIndex);
		if (newIndex == index) return false;
	}
	auto newFrame = std::make_unique<Frame>(*frame, isBlank);
	newFrame->setName("");
	newFrame->setDuration(frame->getDuration() + frame->getStartFrame() - frameIndex);
	frame->setDuration(frameIndex - frame->getStartFrame());
	newFrame->setStartFrame(frameIndex);
	this->frames.insert(this->frames.begin() + index + 1, std::move(newFrame));
	return true;
}

bool Layer::insertKeyframe(unsigned int frameIndex) noexcept {
	return this->insertKeyframe(frameIndex, false);
}

bool Layer::insertBlankKeyframe(unsigned int frameIndex) noexcept {
	return this->insertKeyframe(frameIndex, true);
}

Frame* Layer::getKeyFrame(unsigned int index) const noexcept {
	return frames[index].get();
}

constexpr unsigned int Layer::getKeyframeIndex(unsigned int frameIndex) const noexcept {
	// return the nth keyframe where n.startFrame <= frameIndex < n.startFrame + n.duration using binary search
	unsigned int index = 0;
	unsigned int start = 0;
	unsigned int end = this->frames.size() - 1;
	while (start <= end) {
		unsigned int mid = (start + end) / 2;
		Frame* frame = this->frames[mid].get();
		if (frame->getStartFrame() <= frameIndex && frameIndex < frame->getStartFrame() + frame->getDuration()) {
			index = mid;
			break;
		}
		else if (frame->getStartFrame() > frameIndex) {
			end = mid - 1;
		}
		else {
			start = mid + 1;
		}
	}
	return index;
}

Frame* Layer::getFrame(unsigned int frameIndex) const noexcept {
	unsigned int index = this->getKeyframeIndex(frameIndex);
	return this->frames[index].get();
}

const std::string& Layer::getColor() const noexcept {
	return this->color;
}
void Layer::setColor(const std::string& color) noexcept {
	if (this->root.attribute("color").empty()) this->root.append_attribute("color");
	this->root.attribute("color").set_value(color.c_str());
	this->color = color;
}
const std::string& Layer::getLayerType() const noexcept {
	return this->layerType;
}
void Layer::setLayerType(const std::string& layerType) noexcept(false) {
	if (std::find(ACCEPTABLE_LAYER_TYPES.begin(), ACCEPTABLE_LAYER_TYPES.end(), layerType) == ACCEPTABLE_LAYER_TYPES.end()) {
		throw std::invalid_argument("Invalid layer type specified: " + layerType);
	}
	// normal doesn't show up in the xml
	if (layerType.find("normal") != std::string::npos) this->root.remove_attribute("layerType");
	else {
		if (this->root.attribute("layerType").empty()) this->root.append_attribute("layerType");
		this->root.attribute("layerType").set_value(layerType.c_str());
	}
	this->layerType = layerType;
}
bool Layer::isLocked() const noexcept {
	return this->locked;
}
void Layer::setLocked(bool locked) noexcept {
	if (!locked) this->root.remove_attribute("locked");
	else {
		if (this->root.attribute("locked").empty()) this->root.append_attribute("locked");
		this->root.attribute("locked").set_value(locked);
	}
	this->locked = locked;
}
const std::string& Layer::getName() const noexcept {
	return this->name;
}
void Layer::setName(const std::string& name) noexcept {
	if (this->root.attribute("name").empty()) this->root.append_attribute("name");
	this->root.attribute("name").set_value(name.c_str());
	this->name = name;
}
std::optional<unsigned int> Layer::getParentLayerIndex() const noexcept {
	return this->parentLayerIndex;
}
void Layer::setParentLayerIndex(std::optional<unsigned int> parentLayerIndex) noexcept {
	if (!parentLayerIndex.has_value()) this->root.remove_attribute("parentLayerIndex");
	else {
		if (this->root.attribute("parentLayerIndex").empty()) this->root.append_attribute("parentLayerIndex");
		this->root.attribute("parentLayerIndex").set_value(parentLayerIndex.value());
	}
	this->parentLayerIndex = parentLayerIndex;
}

unsigned int Layer::getFrameCount() const noexcept {
	return this->frames[this->frames.size() - 1]->getStartFrame() + this->frames[this->frames.size() - 1]->getDuration();
}
pugi::xml_node& Layer::getRoot() noexcept {
	return this->root;
}
const pugi::xml_node& Layer::getRoot() const noexcept {
	return this->root;
}

File: CsXFL/Layer.cs
using System.Xml.Linq;

public class Layer
{
    private static readonly List<string> AcceptableLayerTypes = new List<string> { "normal", "guide", "guided", "mask", "masked", "folder", "camera" };
    public static class DefaultValues
    {
        public const string Color = "#000000";
        public const string LayerType = "normal";
        public const string Name = "layer";
        public const bool Locked = false;
        public const bool Current = false;
        public const bool Selected = false;
    }
    private XElement? root;
    private XNamespace ns;
    private string color, layerType, name;
    private List<Frame> frames;
    private bool locked, current, selected;
    uint? parentLayerIndex;
    private void SetOrRemoveAttribute(in string attributeName, object? value, object? defaultValue)
    {
        if (value == defaultValue)
        {
            root?.Attribute(attributeName)?.Remove();
        }
        else
        {
            root?.SetAttributeValue(attributeName, value);
        }
    }
    public XElement? Root { get { return root; } }
    public string Color { get { return color; } set { color = value; root?.SetAttributeValue("color", value); } }
    public string LayerType
    {
        get { return layerType; }
        set
        {
            if (!AcceptableLayerTypes.Contains(value))
            {
                throw new ArgumentException("Invalid layer type: " + value);
            }
            layerType = value;
            SetOrRemoveAttribute("layerType", value, DefaultValues.LayerType);
        }
    }
    public string Name { get { return name; } set { name = value; root?.SetAttributeValue("name", value); } }
    public bool Locked { get { return locked; } set { locked = value; SetOrRemoveAttribute("locked", value, DefaultValues.Locked); } }
    public bool Current { get { return current; } set { current = value; SetOrRemoveAttribute("current", value, DefaultValues.Current); } }
    public bool Selected { get { return selected; } set { selected = value; SetOrRemoveAttribute("isSelected", value, DefaultValues.Selected); } }
    public uint? ParentLayerIndex { get { return parentLayerIndex; } set { parentLayerIndex = value; SetOrRemoveAttribute("parentLayerIndex", value, null); } }

    private void LoadFrames(XElement layerNode)
    {
        List<XElement>? frameNodes = layerNode?.Element(ns + "frames")?.Elements().ToList();
        if (frameNodes is null) return;
        foreach (XElement frameNode in frameNodes)
        {
            frames.Add(new Frame(frameNode));
        }
    }
    public Layer(XElement layerNode)
    {
        root = layerNode;
        ns = root.Name.Namespace;
        color = (string?)layerNode.Attribute("color") ?? DefaultValues.Color;
        layerType = (string?)layerNode.Attribute("layerType") ?? DefaultValues.LayerType;
        if(!AcceptableLayerTypes.Contains(layerType))
        {
            throw new ArgumentException("Invalid layer type: " + layerType);
        }
        name = (string?)layerNode.Attribute("name") ?? DefaultValues.Name;
        locked = (bool?)layerNode.Attribute("locked") ?? DefaultValues.Locked;
        current = (bool?)layerNode.Attribute("current") ?? DefaultValues.Current;
        selected = (bool?)layerNode.Attribute("isSelected") ?? DefaultValues.Selected;
        parentLayerIndex = (uint?)layerNode.Attribute("parentLayerIndex");
        frames = new List<Frame>();
        LoadFrames(root);
    }

    public Layer(ref Layer other)
    {
        root = other.root is null ? null : new XElement(other.root);
        ns = other.ns;
        color = other.color;
        layerType = other.layerType;
        name = other.name;
        locked = other.locked;
        current = other.current;
        selected = other.selected;
        parentLayerIndex = other.parentLayerIndex;
        frames = new List<Frame>();
        if (root is not null) LoadFrames(root);
    }
    private void RemoveKeyframe(int keyframeIndex)
    {
        root?.Element(ns + "frames")?.Elements().ToList()[keyframeIndex].Remove();
        frames.RemoveAt(keyframeIndex);
    }
    int GetKeyframeIndex(int frameIndex)
    {
        // return the nth keyframe where n.StartFrame <= frameIndex < (n.StartFrame + n.Duration) with binary search
        int left = 0, right = frames.Count- 1;   
        while (left <= right)
        {
            int mid = (left + right) / 2;
            Frame midFrame = frames[mid];
            if (midFrame.StartFrame <= frameIndex && frameIndex < (midFrame.StartFrame + midFrame.Duration))
            {
                return mid;
            }
            else if (midFrame.StartFrame > frameIndex)
            {
                right = mid - 1;
            }
            else
            {
                left = mid + 1;
            }
        }
        throw new Exception("Keyframe not found");
    }
    public int GetFrameCount()
    {
        return frames.Count > 0 ? frames[^1].StartFrame + frames[^1].Duration : 0;
    }
    public Frame GetFrame(int frameIndex)
    {
        return frames[GetKeyframeIndex(frameIndex)];
    }
    public bool InsertKeyframe(int frameIndex, bool isBlank)
    {
        int index = GetKeyframeIndex(frameIndex);
        Frame frame = frames[index];
        if(frameIndex == frame.StartFrame)
        {
            frameIndex++;
            if(frameIndex >= GetFrameCount()) return false;
            int newIndex = GetKeyframeIndex(frameIndex);
            if(newIndex != index) return false;
        }
        Frame newFrame = new Frame(ref frame, isBlank)
        {
            Name = Frame.DefaultValues.Name,
            Duration = frame.Duration + frame.StartFrame - frameIndex
        };
        frame.Duration = frameIndex - frame.StartFrame;
        newFrame.StartFrame = frameIndex;
        frames.Insert(index + 1, newFrame);
        frame.Root!.AddAfterSelf(newFrame.Root);
        return true;
    }
}